### PR TITLE
`image2` example - kernel debug/changes

### PR DESCRIPTION
**Pull request by c0gent**
Changes: 

- [minor] Use the `find_folder` crate and `Program::builder().src_file(...)` to load the kernel file more robustly and eliminate the need for the `read_kernel` function.
- Fix the segfault inducing overflow in the global work size calculation and create two separate kernels "patches" and "unrolled". Notes listed at the top of `main.rs` describing why.

The original workgroup size calculation was incorrect, and was accessing memory beyond the bounds of the image buffer, causing segfaults (when the CPU was being used). GPUs usually don't care if you read or write to invalid memory and are extremely forgiving of overruns but because you're using integrated graphics (which counts as a GPU as far as OpenCL is concerned), and therefore system memory, your system segfaulted when mine didn't at first (even when I ran it on my CPU several times -- just due to the randomness of how the memory was allocated). If you had run this on an external GPU device this bug may have gone unnoticed by any of us for who knows how long.

### CODE REVIEW

**Comment by c0gent** [comment]
@drahnr:

I just made some additions and corrections. Let me know how that runs for you.

**Comment by drahnr** [comment]
I disagree.

The original implementation was causing no out of memory access, since the sampler is supposed to clamp the coordinates to the image buffer limits. At least that is what I think I read from the OpenCL spec on the Khronos website....

The idea of adding `((x + patch_size - 1)/patch_size)` was to make sure it rounds up to make sure the whole image is covered.

I am aware that the width and height are not multiples of `patch_size` will cause an out of bounds access _but_ that is why there is a sampler isn't it?

**Comment by c0gent** [comment]
The sampler clamps the values of individual pixels, not the indexes of them. Unfortunately there's no built-in index overrun protection in OpenCL.

**Comment by c0gent** [comment]
On a side note, overrun protection is left out for good reason, bounds checks are expensive (a branch) and would significantly slow things down on a GPU. Looping and branching are to be avoided unless absolutely necessary when processing in parallel because it prevents the OpenCL runtime from optimizing memory reads/writes optimally.

**Comment by drahnr** [comment]
>  Specifies the image addressing-mode i.e. how out-of-range image coordinates are handled. This must be a literal value and can be one of the following predefined enums:
> 
> CLK_ADDRESS_REPEAT - out-of-range image coordinates are wrapped to the valid range. This address mode can only be used with normalized coordinates. If normalized coordinates are not used, this addressing mode may generate image coordinates that are undefined.
> 
> CLK_ADDRESS_CLAMP_TO_EDGE - out-of-range image coordinates are clamped to the extent.
> 
> CLK_ADDRESS_CLAMP32 - out-of-range image coordinates will return a border color. The border color is (0.0f, 0.0f, 0.0f, 0.0f) if image channel order is CL_A, CL_INTENSITY, CL_RA, CL_ARGB, CL_BGRA or CL_RGBA and is (0.0f, 0.0f, 0.0f, 1.0f) if image channel order is CL_R, CL_RG, CL_RGB or CL_LUMINANCE.
> 
> CLK_ADDRESS_NONE - for this address mode the programmer guarantees that the image coordinates used to sample elements of the image refer to a location inside the image; otherwise the results are undefined. 
https://www.khronos.org/registry/cl/sdk/1.0/docs/man/xhtml/sampler_t.html

**Comment by drahnr** [comment]
I think it is exactly what I expected or did it change significantly between 1.0 and 1.2?

**Comment by drahnr** [comment]
Or did I get this entirely wrong?

**Comment by c0gent** [comment]
No you just forgot to set your sampler to normalize coords :)

**Comment by c0gent** [comment]
```
__constant sampler_t sampler_const =
CLK_NORMALIZED_COORDS_FALSE |
CLK_ADDRESS_CLAMP |
CLK_FILTER_NEAREST;
```

**Comment by c0gent** [comment]
It's a big performance boost to leave Normalized coords _off_ by the way.

**Comment by c0gent** [comment]
Just to clarify and expand on what I meant, clamping the image coordinates is generally intended to work only with normalized coordinates. Most vendors (including Intel) don't implement the clamping when indexes are specified using an integer. This is for performance reasons. This is another area where the behavior on certain platforms doesn't always mesh exactly with what the documentation says.

**Comment by drahnr** [comment]
But the reference to normalized coords is only for REPEAT. At least that's how I read it...

**Comment by c0gent** [comment]
You are entirely correct. The spec is just not always implemented properly and must be taken with a grain of salt.  Lots and lots of instances of this that I've had to learn the hard way and take into account (much of this is baked into the library and I expect to find plenty more cases in the future). 

Some vendors like AMD do a great job of implementing the spec precisely... others don't. Usually Intel is very good about these things but as I mentioned, I think in this case the current behavior is more of a performance optimization than anything. 

Don't worry... you're not the first to be frustrated by it. Your reasoning is perfectly sound.

You may want to experiment with `CLK_ADDRESS_CLAMP_TO_EDGE` or some of the other modes if you really need the normalization. Keep in mind, though, that leaving index normalization to the driver will always incur a performance cost  (for both integer and float index values).  It's always best practice to define your bounds precisely, especially when dealing with the compute side of things because there is no normalization or safety net there.

If you approve, I'll merge this PR. Please feel free to make any changes or corrections you'd like or let me know if you have any other questions / criticisms / arguments / ideas.

**Comment by drahnr** [comment]
Sigh. Ok, in that case, I guess the adjustments you made are totally sound.

**Comment by c0gent** [comment]
Ok let me know if you want me to merge this for now or if you'd rather do something else with your example.

**Comment by drahnr** [comment]
I think documenting these kind of nuances and common deviations and performance implications should be added to the example itself.

Otherwise I am good with the adaptations you made.

**Comment by drahnr** [comment]
Actually, wait.

I tried it again with the finish, it is not reproduceable what I've seen before.

I also tried CPU only and I did not get segfaults.

So I'd like to run a few more tests within the next few days before merging this.

**Comment by c0gent** [comment]
Doh, I just noticed your latest comment as I was merging. Just sumbit another PR with any other changes you want to make.

**Comment by c0gent** [comment]
As to whether or not you're getting segfaults... You won't *always* get them, even if out of bounds indexing is taking place. You won't get them at all on some platforms/devices. This is just due to randomness in how buffer memory is allocated within host memory and the fact that if buffer memory is allocated on a device with dedicated memory, it will read/write to anywhere without concern.

Getting the segfaults at all is a happy circumstance that indicates there is an indexing problem.

Basically the only way to be sure is to do the math correctly and carefully, just like any other kind of indexing.

**Comment by c0gent** [comment]
> Basically the only way to be sure is to do the math correctly and carefully, just like any other kind of indexing.

Just to expand on this a bit: it isn't entirely true. There are ways to manually clamp indexed buffer values  even without using branching. There is a very small performance cost but nothing near as bad as a branching bounds check. If you're interested in knowing more I can point you to some examples and give you more info.

**Comment by drahnr** [comment]
A link would be really appreciated :)

**Comment by drahnr** [comment]
If you get those kind of errors where I do not, which platform are you on?

**Comment by c0gent** [comment]
> If you get segfaults where I do not, which platform are you on?

I usually only get the segfaults when I run on Intel(CPU)/Linux and even then only sporadically. Intel/Windows will occasionally do the same. Again, this is a completely terrible way to evaluate kernels for correctness, though I admit that I rely on it way more than I should. There are lots of OpenCL tools to diagnose potential bounds problems such as [AMDs CodeXL](http://developer.amd.com/tools-and-sdks/opencl-zone/codexl/) which I have limited knowledge of and I guess has been renamed. With minimal learning you could probably teach me plenty about this topic.

> A link would be really appreciated :)

Here is a quick (and admittedly obscure) example from one of my projects: 
`coord_is_safe()`: https://github.com/cogciprocate/bismit/blob/master/cl/bismit.cl#L252
`axn_idx_3d_unsafe()`: https://github.com/cogciprocate/bismit/blob/master/cl/bismit.cl#L318
`axn_state_3d_safe()`: https://github.com/cogciprocate/bismit/blob/master/cl/bismit.cl#L392

I'll provide you a simplified example and more explanation tomorrow but the gist of it is that you can do cheap bounds checks by simply using a few bitwise and comparison operators, such as `(index >= index_lower_bound) & (index < index_upper_bound)` along with a couple of multiply operations to get a "safe" value without branching.

**Comment by c0gent** [comment]
I'm guessing you're European due to your activity times. I'm from California but I should be awake for several more hours working on a project so if you prod me I can elaborate further.

**Comment by drahnr** [comment]
Munich to be exact :)

It for sure is not hard, but still quite some bloat give that `sampler_t` was meant for this purpose.
```
const uint32_t b _lower = (index <= 0);
const uint32_t b _lower = (index >= width);
clamped_x = ((1-b_lower) & (1-b_upper)) * x + b_upper * (width-1) + b_lower * 0;
```

I read about CodeXL a few months ago or so but never got around to actually checking it out, but that is definitely on my list for this month.

**Comment by c0gent** [comment]
You've already got it figured out :)

> Thanks a million for your time and feedback.

My pleasure. Thank you for your time and feedback as well.
